### PR TITLE
nodes auto-discovery and accessing on private IPs

### PR DESCRIPTION
**Pull request by akostadinov**
first it is not necessary to specify all nodes in host spec.
nodes with missing hosts will try to use node hostname from the Node object.
By default as bastion host will be used the firsts master. But if you
set a ole to some host as bastion - it will be used instead.

So you can also have completely internal cluster if you setup a special
bastion host.

The main idea though is that you can directly feed host.spec from
flexy into jenkins and have everything working automatically.

HTH

### CODE REVIEW

**Comment by lihongan** [comment]
LGTM, thanks

**Comment by pruan-rht** [comment]
LGTM

File: lib/host.rb [hunk 1]
```diff
@@ -759,6 +778,42 @@ def copy_from(remote, local, **opts)
...
+      end
+    end
+
+    private def select_bastion(hosts)
+      hosts = hosts.select { |h| self.class === h }
+      b = hosts.find {|h| h.roles.include?(:bastion)}
+      unless b
+        # select first master as a bastion host
+        b = hosts.find { |h|
+          h.roles.include?(:master) && !h[:flags].include?("/b/")
```
Comment (xingxingxia) [inline lib/host.rb]: @akostadinov 
With :bastion set, `When I run commands on the host:` passed: https://openshift-qe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/Runner-v3/37541/console 
Without :bastion set, it failed: https://openshift-qe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/Runner-v3/37546/console :
```
      undefined method `include?' for nil:NilClass (NoMethodError)
      /home/jenkins/workspace/Runner-v3/lib/host.rb:810:in `block in select_bastion'
      /home/jenkins/workspace/Runner-v3/lib/host.rb:809:in `each'
      /home/jenkins/workspace/Runner-v3/lib/host.rb:809:in `find'
      /home/jenkins/workspace/Runner-v3/lib/host.rb:809:in `select_bastion'
      /home/jenkins/workspace/Runner-v3/lib/host.rb:784:in `apply_flags'
      /home/jenkins/workspace/Runner-v3/lib/environment.rb:433:in `host_add'
      /home/jenkins/workspace/Runner-v3/lib/openshift/node.rb:51:in `host'
      /home/jenkins/workspace/Runner-v3/features/step_definitions/node.rb:13:in `/^I select a random node's host$/'
      features/tierN/cli/service.feature:82:in `And I select a random node's host'
```

Comment (akostadinov) [reply lib/host.rb]: fixed with #71

**Pull request merged**